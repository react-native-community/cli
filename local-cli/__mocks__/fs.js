/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

'use strict';

const path = require('path');
const MemoryFS = require('metro-memory-fs');
let fs;

function setMockFilesystem(object, platform) {
  reset(platform);
  const root = platform === 'win32' ? 'c:\\' : '/';
  mockDir(root, {...object});
  return root;
}

function mockDir(dirPath, desc) {
  for (const entName in desc) {
    const ent = desc[entName];
    const entPath = path.join(dirPath, entName);
    if (typeof ent === 'string' || ent instanceof Buffer) {
      fs.writeFileSync(entPath, ent);
      continue;
    }
    if (typeof ent !== 'object') {
      throw new Error(require('util').format('invalid entity:', ent));
    }
    if (ent.SYMLINK != null) {
      fs.symlinkSync(ent.SYMLINK, entPath);
      continue;
    }
    fs.mkdirSync(entPath);
    mockDir(entPath, ent);
  }
}

function reset(platform) {
  if (path.mock == null) {
    throw new Error(
      'to use this "fs" module mock, you must also mock the "path" module',
    );
  }
  path.mock.reset(platform);
  const cwd = () => (platform === 'win32' ? 'c:\\' : '/');
  fs = new MemoryFS({platform, cwd});
  Object.assign(mockFs, fs);
}

const mockFs = {};
mockFs.__setMockFilesystem = setMockFilesystem;
mockFs.mock = {clear: reset};

reset('posix');

module.exports = mockFs;
