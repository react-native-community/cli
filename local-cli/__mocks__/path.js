/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

'use strict';

const mockPath = {};

function reset(platform) {
  Object.assign(mockPath, jest.requireActual('path')[platform]);
}

mockPath.mock = {reset};

reset('posix');

module.exports = mockPath;
