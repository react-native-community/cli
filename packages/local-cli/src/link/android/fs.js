/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

const fs = require('fs-extra');

exports.readFile = file => () => fs.readFileSync(file, 'utf8');

exports.writeFile = (file, content) =>
  content
    ? fs.writeFileSync(file, content, 'utf8')
    : c => fs.writeFileSync(file, c, 'utf8');
