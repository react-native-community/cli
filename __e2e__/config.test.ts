import path from 'path';
import fs from 'fs';
import {wrap} from 'jest-snapshot-serializer-raw';
import {
  runCLI,
  getTempDirectory,
  cleanupSync,
  writeFiles,
  spawnScript,
  replaceProjectRootInOutput,
} from '../jest/helpers';

const DIR = getTempDirectory('test_root');

function isValidJSON(text: string) {
  try {
    JSON.parse(text);
    return true;
  } catch {
    return false;
  }
}

// We have to check whether setup_env script fails, if it does then we shouldn't log any info to the console
function createCorruptedSetupEnvScript() {
  const originalSetupEnvPath = path.join(
    __dirname,
    '../packages/cli/setup_env.sh',
  );
  const originalSetupEnv = fs.readFileSync(originalSetupEnvPath);
  const corruptedScript = '#!/bin/sh\n exit 1;';
  fs.writeFileSync(originalSetupEnvPath, corruptedScript);
  return () => {
    fs.writeFileSync(originalSetupEnvPath, originalSetupEnv);
  };
}

beforeAll(() => {
  // Register all packages to be linked
  for (const pkg of ['platform-ios', 'platform-android']) {
    spawnScript('yarn', ['link'], {
      cwd: path.join(__dirname, `../packages/${pkg}`),
    });
  }

  // Clean up folder and re-create a new project
  cleanupSync(DIR);
  writeFiles(DIR, {});

  // Initialise React Native project

  runCLI(DIR, ['init', 'TestProject']);

  // Link CLI to the project
  const pkgs = [
    '@react-native-community/cli-platform-ios',
    '@react-native-community/cli-platform-android',
  ];

  spawnScript('yarn', ['link', ...pkgs], {
    cwd: path.join(DIR, 'TestProject'),
  });
});

afterAll(() => {
  cleanupSync(DIR);
});

test('shows up current config without unnecessary output', () => {
  const {stdout} = runCLI(path.join(DIR, 'TestProject'), ['config']);
  const parsedStdout = JSON.parse(stdout);
  // Strip unnecessary parts
  parsedStdout.commands = parsedStdout.commands.map((command: any) => ({
    ...command,
    examples: command.examples && ['<<REPLACED>>'],
    options: command.options && ['<<REPLACED>>'],
  }));

  const configWithReplacedProjectRoots = replaceProjectRootInOutput(
    JSON.stringify(parsedStdout, null, 2).replace(/\\\\/g, '\\'),
    DIR,
  );
  expect(wrap(configWithReplacedProjectRoots)).toMatchSnapshot();
});

test('should log only valid JSON config if setting up env throws an error', () => {
  const restoreOriginalSetupEnvScript = createCorruptedSetupEnvScript();
  const {stdout, stderr} = runCLI(path.join(DIR, 'TestProject'), ['config']);

  restoreOriginalSetupEnvScript();
  expect(isValidJSON(stdout)).toBe(true);
  expect(stderr).toBe('');
});
