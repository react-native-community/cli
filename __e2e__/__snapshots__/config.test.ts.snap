// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows up current config without unnecessary output 1`] = `
{
  "root": "<<REPLACED_ROOT>>/TestProject",
  "reactNativePath": "<<REPLACED_ROOT>>/TestProject/node_modules/react-native",
  "dependencies": {},
  "commands": [
    {
      "name": "log-ios",
      "description": "starts iOS device syslog tail"
    },
    {
      "name": "run-ios",
      "description": "builds your app and starts it on iOS simulator",
      "examples": [
        "<<REPLACED>>"
      ],
      "options": [
        "<<REPLACED>>"
      ]
    },
    {
      "name": "log-android",
      "description": "starts logkitty"
    },
    {
      "name": "run-android",
      "description": "builds your app and starts it on a connected Android emulator or device",
      "options": [
        "<<REPLACED>>"
      ]
    }
  ],
  "assets": [],
  "platforms": {
    "ios": {},
    "android": {}
  },
  "project": {
    "ios": {
      "sourceDir": "<<REPLACED_ROOT>>/TestProject/ios",
      "folder": "<<REPLACED_ROOT>>/TestProject",
      "pbxprojPath": "<<REPLACED_ROOT>>/TestProject/ios/TestProject.xcodeproj/project.pbxproj",
      "podfile": "<<REPLACED_ROOT>>/TestProject/ios/Podfile",
      "podspecPath": null,
      "projectPath": "<<REPLACED_ROOT>>/TestProject/ios/TestProject.xcodeproj",
      "projectName": "TestProject.xcodeproj",
      "libraryFolder": "Libraries",
      "sharedLibraries": [],
      "plist": [],
      "scriptPhases": []
    },
    "android": {
      "sourceDir": "<<REPLACED_ROOT>>/TestProject/android",
      "isFlat": true,
      "folder": "<<REPLACED_ROOT>>/TestProject",
      "stringsPath": "<<REPLACED_ROOT>>/TestProject/android/app/src/main/res/values/strings.xml",
      "manifestPath": "<<REPLACED_ROOT>>/TestProject/android/app/src/main/AndroidManifest.xml",
      "buildGradlePath": "<<REPLACED_ROOT>>/TestProject/android/build.gradle",
      "settingsGradlePath": "<<REPLACED_ROOT>>/TestProject/android/settings.gradle",
      "assetsPath": "<<REPLACED_ROOT>>/TestProject/android/app/src/main/assets",
      "mainFilePath": "<<REPLACED_ROOT>>/TestProject/android/app/src/main/java/com/testproject/MainApplication.java",
      "packageName": "com.testproject",
      "packageFolder": "com/testproject",
      "appName": "app"
    }
  }
}
`;
